File: problems/problemset/chess-engine/C++/runner/Testbench.cpp
#include <bits/stdc++.h>
#include <nlohmann/json.hpp>
#include <chrono>
#include <sys/resource.h>
#include "../code/Chess.h"

using json = nlohmann::json;

static long peakMemoryKB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}

int main() {
    std::string testcase_str((std::istreambuf_iterator<char>(std::cin)),
                              std::istreambuf_iterator<char>());

    json testcase = json::parse(testcase_str);
    auto input = testcase["input"].get<std::string>();

    std::stringstream ss(input);
    std::string line;
    std::string output;
    Chess game;

    auto start = std::chrono::high_resolution_clock::now();

    while (std::getline(ss, line)) {
        if (line.empty()) continue;
        std::istringstream ls(line);
        std::string cmd;
        ls >> cmd;

        if (cmd == "NEW") {
            game.newGame();
        } else if (cmd == "MOVE") {
            std::string mv;
            ls >> mv;
            output += game.move(mv) + '\n';
        } else if (cmd == "PRINT") {
            output += game.print();
        } else if (cmd == "MOVES") {
            std::string sq;
            ls >> sq;
            output += game.moves(sq) + '\n';
        } else if (cmd == "STATUS") {
            output += game.status() + '\n';
        } else if (cmd == "HISTORY") {
            output += game.history();
        }
    }

    auto end = std::chrono::high_resolution_clock::now();
    double time_ms   = std::chrono::duration<double, std::milli>(end - start).count();
    long   memory_kb = peakMemoryKB();

    json result;
    result["actual_output"] = output;
    result["time_ms"]       = time_ms;
    result["memory_kb"]     = memory_kb;

    std::cout << "__BEGIN_RESULT__" << result.dump() << "__END_RESULT__";
}

File: problems/problemset/chess-engine/Python/runner/testbench.py
import sys
import json
import time
import resource

sys.path.insert(0, '/workspace/code')
from chess import Chess

def main():
    testcase = json.loads(sys.stdin.read())
    game = Chess()
    output_lines = []

    start = time.perf_counter()

    for line in testcase['input'].splitlines():
        parts = line.split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == 'NEW':
            game.new_game()
        elif cmd == 'MOVE':
            output_lines.append(game.move(parts[1]))
        elif cmd == 'PRINT':
            output_lines.append(game.print_board().rstrip('\n'))
        elif cmd == 'MOVES':
            sq = parts[1] if len(parts) > 1 else ''
            output_lines.append(game.moves(sq))
        elif cmd == 'STATUS':
            output_lines.append(game.status())
        elif cmd == 'HISTORY':
            output_lines.append(game.history().rstrip('\n'))

    elapsed_ms = (time.perf_counter() - start) * 1000
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    actual_output = '\n'.join(output_lines) + '\n' if output_lines else ''

    result = {'actual_output': actual_output, 'time_ms': elapsed_ms, 'memory_kb': memory_kb}
    sys.stdout.write(f'__BEGIN_RESULT__{json.dumps(result)}__END_RESULT__')

main()
